### ISSUE TITLE
Simplifying lifetimes across different Types and Traits

### ISSUE THREAD

**Issue by cmaves** [KillingSpark/rustbus#60]
IMO there are a number of places in this crate where there are unnecessary or needlessly complex lifetimes that hurt the ergonomics of the crate. Personally I find that lifetimes are one of Rust's most difficult concepts that creates the most issues.
Reducing/simplifying them can reduce the cognitive load they create, make code more intuitive and prettier.

To give an example of what I mean take the `Unmarshal` trait and its related types:
```rust
pub trait Unmarshal<'r, 'buf: 'r, 'fds>: Sized + Signature {
    fn unmarshal(ctx: &mut UnmarshalContext<'fds, 'buf>) -> unmarshal::UnmarshalResult<Self>;
}
pub struct UnmarshalContext<'fds, 'buf> {
    pub fds: &'fds [crate::wire::UnixFd],
    pub buf: &'buf [u8],
    pub byteorder: ByteOrder,
    pub offset: usize,
}
```
In `UnmarshalContext` is only created `MarshalledMessageBody`. This makes the `fds and `buf always equivalent because they come from the same borrow of the body, so simplify to: 
```rust
pub trait Unmarshal<'r, 'buf: 'r>: Sized + Signature {
    fn unmarshal(ctx: &mut UnmarshalContext<'buf>) -> unmarshal::UnmarshalResult<Self>;
}
pub struct UnmarshalContext<'buf> {
    pub fds: &'buf [crate::wire::UnixFd],
    pub buf: &'buf [u8],
    pub byteorder: ByteOrder,
    pub offset: usize,
}
```
We can go further with the `Unmarshal` trait. Because lifetimes are transitive across nested immutable borrows, the type returned from `unmarshal()` should is equivelent to `'buf`. This makes `'r` unnecessary and can be eliminated:
```rust
pub trait Unmarshal<'buf>: Sized + Signature {
    fn unmarshal(ctx: &mut UnmarshalContext<'buf>) -> unmarshal::UnmarshalResult<Self>;
}
```
PR #61 demonstrates this particular change. There are other examples that should also probably be looked at.

**Comment by cmaves** [comment]
I'm open to hearing if there is something that I have overlooked with this issue or the above PR, but I think it would ultimately make the library easier to use.

Because it is a significant breaking change that doesn't add features or performance, that is a low priority issue that should probably be resolved carefully (if at all).

**Comment by KillingSpark** [comment]
Those are good questions! I do not have a definitive answer right now, for why I did this. I remember how much time I spent refactoring all of this code to use the more complex lifetimes but I do not remember the exact reason why. I am pretty sure I had some example where it did not work. That might have been just an issue with rusts lifetime resolution that has been fixed in newer releases too... I'll think about it. I am somewhat sure it had to do with `&str` unmarshalling.

It definitely is something that was carried over from when no Marshal/UnmarshalContext was used. Maybe it made more sense then to have separate `'buf` and `'fds` lifetimes.

In the end this is also one of the first times I really messed with more complex lifetime scenarios, so it is entirely possible that those are just needlessly complex because I stopped fiddling around once it worked.

**Comment by KillingSpark** [comment]
Ok so I came up with a reason to have at least `'buf` and `'fds` be separate lifetimes at least for the trait definition. If 'fds is significantly smaller than 'buf (which is never the case in this crate but the trait should be usable outside of this crate too) then the result would be unncessarily tied to the smaller lifetime. The 'fds lifetime is more or less irrelevant for the returned value, but it is important to tell the compiler that this is the case. 

I agree that the 'r lifetime is very likely not needed since the returned values are always valid for at least as long `'buf`. 

```rust
fn main() {
    let mut buffer = [0u8,1,2,3,4,5];
    // works because foo does not depend on the lifetime of "fds"
    let foo = {
        let fds = [0u8,1,2,3,4,5,6,7,8];
        foo(&buffer, &fds)
    };
    // does not work because bar does depend on the lifetime of "fds"
    let bar = {
        let fds = [0u8,1,2,3,4,5,6,7,8];
        bar(&buffer, &fds)
    };
    println!("{}", foo[0]);
    println!("{}", bar[0]);
    buffer[0] = 10;
}

fn foo<'r, 'buf: 'r, 'fds>(buf: &'buf [u8], fds: &'fds [u8]) -> &'r [u8] {
    let _ = fds[0];
    return &buf[0..1];
}
fn bar<'r, 'buf: 'r>(buf: &'buf [u8], fds: &'buf [u8]) -> &'r [u8] {
    let _ = fds[0];
    return &buf[0..1];
}
```

**Comment by cmaves** [comment]
I see what you getting at but I don't think it reflect a real world use case, because of how tightly linked `&[UnixFd]` and `&[u8]` are to each other in practice. 

There are two possible scenarios to consider:
* **'buf outlives 'fds** (like in the example above): This is almost never valid, because `buf` encodes indices within the `fds`. Generally, `buf` should only be valid within the context of the `fds` slice.
* **'fds outlives 'buf**:  While this is more likely to be valid and more do to `fds` generally being smaller, it probably isn't common because you receive both the fds and buffer from the OS at the same time so they start with same lifetime. 
  * If you only care about a single `&UnixFd`, just clone it. 
  * If you only care about the `fds` in a message (and wish to discard the ordering/context provided by the `buf`), unmarshalling is probably an inefficient way to get them, because you already have access to the buffer of `fds` you used to create `UnmarshalContext` (in a use case external to rustbus). 
  * If you want to get the `fds` in "message-order" list, you could return `Vec<&UnixFd>`, but it would cheap and more flexible to clone it into `Vec<UnixFd>`.  The same principle applies to other data structures you might return.

The tl;dr is that `UnixFd` is a reference counted type, which kinda negates the the need for borrowed references.

**Comment by KillingSpark** [comment]
My argument would be that it is especially valuable **because** fd is ref counted. You can unmarshal and then drop the `Vec<UnixFD>` but the fds are still in the unmarshalled thing. That would not be possible if the unmarshalled thing is bound to the  `Vec<UnixFD>`'s lifetime.

I know that isn't really relevant to this concrete lib but I'd like to keep this core abstraction as widely applicable as possible.

**Comment by cmaves** [comment]
Ultimately I really getting rid of the `'r` lifetime is my primary concern. Removing `'fds` was something I felt should be done since we already are making a breaking change, but I can revert it that particular aspect.

**Comment by cmaves** [comment]
I got one final issue to discuss with `'fds` before re-adding it. 

The existing `#[derive(Unmarshal)]`(before this issue's PR) is broken for types that return the `'fds` lifetime when unmarshalling. AFAIK, the only way to resolve this problem would be to have `#[derive(Unmarshal)]` produce `impl<'r> Unmarshal<'r, 'r, 'r> for SomeType`. This is because if you are deriving a type that has a reference, you do not know which lifetime (`'fds` or `'ret`) it is associated with.  This problem will persist even when `'r` is removed.

Its problems like this one that motivates my push for `'fds` outright removal. Its existence creates odd lifetime issues for a what is likely a insignificant theoretical performance benefit. Either way it is up to you. I can go and re-add `'fds` or leave it as it is.

**Comment by KillingSpark** [comment]
Yeah I see why having the `'fds` is sometimes problematic. I do still think being more explicit here is better though. I agree that the `'r` should be removed and just be `'buf`.

The `derive` issue is something that didn't occur to me yet. But I guess there is already something wrong if that type has that lifetime? The unmarshalling should in all cases not produce something depending on the lifetime of the fd vector. It's ugly that this is broken, but I don't think it is really an issue. But it should be made clear in the doc for the Unmarshal trait.

I know that is another pitfall created by having the `'fds`.

I get the push for a simpler design here, but lifetimes just are complicated. And modelling them correctly and explicitly is (in my opinion) important. This comes with added complexity in the code. But I'd hate it even more if in the future some usecases just cant be done without reverting this change back.

I am sorry for wasting your efforts, but I'd really like to keep the `'fds`.

**Comment by cmaves** [comment]
The lifetime has been re-added.

**Comment by KillingSpark** [comment]
Thank you!
